src/rules/selector-max-depth/index.js

```diff
@@ -0,0 +1,78 @@
+import resolvedNestedSelector from "postcss-resolve-nested-selector"
+
+import {
+  isStandardRule,
+  isStandardSelector,
+  report,
+  ruleMessages,
+  validateOptions,
+} from "../../utils"
+
+export const ruleName = "selector-max-depth"
+
+export const messages = ruleMessages(ruleName, {
+  expected: (selector, depth) => `Expected "${selector}" to have a depth equal to or less than ${depth}`,
+})
+
+export default function (max) {
+  return (root, result) => {
+    const innerSelectorReg = /(?::not\((.*?)\))/g
+    const validOptions = validateOptions(result, ruleName, {
+      actual: max,
+      possible: [function (max) {
+        return typeof max === "number" && max > 0
+      }],
+    })
+    if (!validOptions) { return }
+
+    // Compares depth of a selector with the maximum
+    function checkDepth(selector, rule) {
```

Comment:
@dryoma Firstly, thanks for porting this rule over. Secondarily, thanks for taking the time to consider and test the performance of the rule! :)

I don't think there's a clear right answer here and you both raise valid points. I agree the docs will need to be updated once we've agreed on this. I think it's worth thrashing this issue out though and coming to a consensus that we can document as it'll help future contributors.

There's one additional data point worth considering; PostCSS will [likely parse selectors](https://github.com/postcss/postcss/issues/754) in `6.0`. Perhaps standardising on `postcss-selector-parser` now will help when we come to migrate to PostCSS `6.0`? 

I guess the core question is; do we prioritise standardising implementation over performance, or the other way around?